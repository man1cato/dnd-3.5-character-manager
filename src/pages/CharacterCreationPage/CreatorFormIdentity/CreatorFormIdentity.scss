@import '../CharacterCreationPage';


.CreatorFormIdentity__group {
   @extend .CreatorForm__group;
   > :first-child {
      width: 35%;
      padding-right: $s-size;
   }
   > :last-child {
      width: 65%;
      flex-wrap: wrap;
   }
}

.CreatorFormIdentity__group--top {
   @extend .CreatorFormIdentity__group;
   align-items: flex-start;
} 

.CreatorFormIdentity__group--bottom {
   @extend .CreatorFormIdentity__group;
   align-items: flex-end;
} 

.CreatorFormIdentity__list {
   margin: 0;
   li {
      margin-left: -20px;
   }
}