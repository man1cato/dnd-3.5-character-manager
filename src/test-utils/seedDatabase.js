import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}