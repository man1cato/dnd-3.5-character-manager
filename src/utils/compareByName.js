export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}