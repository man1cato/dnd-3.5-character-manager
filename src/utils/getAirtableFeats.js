import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}