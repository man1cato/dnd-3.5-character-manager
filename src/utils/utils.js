import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])