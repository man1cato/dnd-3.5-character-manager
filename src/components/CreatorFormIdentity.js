import React, { useEffect }  from 'react'
import { Field, ErrorMessage } from 'formik'
import _ from 'lodash'

import SpecialAbilities from './SpecialAbilities'
import { apiObjectToArray } from '../utils/utils'


const CreatorFormIdentity = ({
	selectedRace, 
	setSelectedRace,
	races, 
	jobClasses,
	handleChange,
	setTouched
}) => {
	useEffect(() => {
		setTouched({ remainingSkillPoints: true, feats: true, equipment: true })
	},[])

	return (
		<>
			<div className="form-group--35">
				<h4>Name:</h4>
				<div>
					<Field className="text-input" name="name" />
					<ErrorMessage className="form-group--error" name="name" component="div"/>
				</div>
			</div>
			
			<div className="form-group--35">
				<h4>Gender:</h4>
				<Field className="select" name="gender" component="select" >
					<option value="Male">Male</option>
					<option value="Female">Female</option>
					<option value="Non-Binary">Non-Binary</option>
				</Field>                                    
			</div>

			<div className="form-group--35">
				<h4>Age:</h4>
				<div>
					<div className="input-group">
						<Field className="text-input" name="age" type="number" />
						<div> yrs</div>
					</div>
					<ErrorMessage className="form-group--error" name="age" component="div"/>
				</div>
			</div>

			<div className="form-group--35">
				<h4>Height:</h4>
				<div>
					<div className="input-group">
						<Field className="text-input" name="height.ft" type="number" />
						<div>ft</div>
						<Field className="text-input" name="height.in"  type="number" />
						<div>in</div>
					</div>
					<div className="input-group">
						<ErrorMessage className="form-group--error" name="height.ft" component="div"/>
						<ErrorMessage className="form-group--error" name="height.in" component="div"/>
					</div>
				</div>
			</div>

			<div className="form-group--35">
				<h4>Weight:</h4>
				<div>
					<div className="input-group">
						<Field className="text-input" name="weight" type="number" />
						<div>lbs</div>
					</div>
					<ErrorMessage className="form-group--error" name="weight" component="div"/>
				</div>
			</div>

			<div className="divider"></div>

			<div className="form-group--35">
				<h4>Race:</h4>
				<Field 
					className="select" 
					name="race" 
					component="select" 
					onChange={(e) => {
						setSelectedRace(races[e.target.value])
						handleChange(e)
					}}
				>
					{_.map(apiObjectToArray(races), (race, i) => (
						<option 
							value={race.id} 
							key={`race${i}`}
						>
							{race.name}
						</option>
					))}
				</Field>                                    
			</div>

			<div className="form-group--35 align-bottom">
				<h4>Favored Class:</h4>
				<div>
					{selectedRace.favoredClass ? 
						jobClasses[selectedRace.favoredClass].name
						:
						'Any'
					}
				</div>
			</div>

			<div className="form-group--35">
				<h4>Size:</h4>                                 
				<div>{selectedRace.size}</div>
			</div> 

			<div className="form-group--35">
				<h4>Speed:</h4>                                 
				<div>{selectedRace.speed}</div>
			</div> 

			<div className="form-group--35 align-bottom">
				<h4>Racial Modifiers:</h4>     
				<div className="input-group flex-wrap">
					{_.filter(selectedRace.abilityMods, (mod) => mod !== 0).length > 0 ?
						_.filter(_.toPairs(selectedRace.abilityMods), (mod) => mod[1] !== 0).map((mod) => (
							<div key={mod}>{mod[0].toUpperCase()}: {mod[1]}</div>
						))
						:
						<div>None</div>
					}            
				</div>
			</div>

			<div className="form-group--35 align-bottom">
				<h4>Default Language(s):</h4>
				<div>{selectedRace.defaultLanguages.join(', ')}</div>
			</div>

			<div className="form-group--35 align-top">
				<h4>Bonus Languages:</h4>
				<div>
					{selectedRace.bonusLanguages.join(', ')}
				</div>                                    
			</div>

			<div className="form-group--35 align-top">
				<h4>Racial Bonuses:</h4>
				{selectedRace.racialBonuses ? 
					<ul className="form-group__list">
						{selectedRace.racialBonuses.map((item, i) => (
							<li key={`racialBonus${i}`}>{item}</li>
						))}
					</ul>
					:
					<p>None</p>
				}
			</div>

			<div className="form-group--35 align-top">
				<h4>Special Abilities:</h4>
				<div>
					{selectedRace.specialAbilities ?
						<SpecialAbilities specialAbilityIds={selectedRace.specialAbilities} />
						:
						"None"
					}
				</div>
			</div>        
			
		</>
	)
}


export default CreatorFormIdentity