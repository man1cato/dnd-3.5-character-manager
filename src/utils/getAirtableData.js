import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}