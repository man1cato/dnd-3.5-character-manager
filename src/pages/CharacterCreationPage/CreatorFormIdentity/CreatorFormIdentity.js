import React, { useEffect }  from 'react'
import { Field, ErrorMessage } from 'formik'
import _ from 'lodash'

import SpecialAbilities from '../../../components/SpecialAbilities/SpecialAbilities'
import { apiObjectToArray } from '../../../utils/utils'
import './CreatorFormIdentity.scss'


const CreatorFormIdentity = ({
	values,
	races, 
	jobClasses,
	setTouched,
	setFieldValue
}) => {
	const { race } = values 
	useEffect(() => {
		setTouched({ remainingSkillPoints: true, feats: true, equipment: true, remainingGold: true })
	},[])

	return (
		<div className="container--body">		
			<div className="CreatorFormIdentity__group">
				<h4>Name:</h4>
				<div>
					<Field className="text-input" name="name" />
					<ErrorMessage className="CreatorForm__error" name="name" component="div"/>
				</div>
			</div>
			
			<div className="CreatorFormIdentity__group">
				<h4>Gender:</h4>
				<Field className="select" name="gender" component="select" >
					<option value="Male">Male</option>
					<option value="Female">Female</option>
					<option value="Non-Binary">Non-Binary</option>
				</Field>                                    
			</div>

			<div className="CreatorFormIdentity__group">
				<h4>Age:</h4>
				<div>
					<div className="CreatorForm__input-group">
						<Field className="text-input" name="age" type="number" />
						<div> yrs</div>
					</div>
					<ErrorMessage className="CreatorForm__error" name="age" component="div"/>
				</div>
			</div>

			<div className="CreatorFormIdentity__group">
				<h4>Height:</h4>
				<div>
					<div className="CreatorForm__input-group">
						<Field className="text-input" name="height.ft" type="number" />
						<div>ft</div>
						<Field className="text-input" name="height.in"  type="number" />
						<div>in</div>
					</div>
					<div className="CreatorForm__input-group">
						<ErrorMessage className="CreatorForm__error" name="height.ft" component="div"/>
						<ErrorMessage className="CreatorForm__error" name="height.in" component="div"/>
					</div>
				</div>
			</div>

			<div className="CreatorFormIdentity__group">
				<h4>Weight:</h4>
				<div>
					<div className="CreatorForm__input-group">
						<Field className="text-input" name="weight" type="number" />
						<div>lbs</div>
					</div>
					<ErrorMessage className="CreatorForm__error" name="weight" component="div"/>
				</div>
			</div>

			<div className="divider"></div>

			<div className="CreatorFormIdentity__group">
				<h4>Race:</h4>
				<select 
					className="select" 
					defaultValue={_.findKey(races, race)}
					onChange={(e) => setFieldValue('race', races[e.target.value])}
				>
					{_.map(apiObjectToArray(races), race => (
						<option 
							value={race.id} 
							key={race.id}
						>
							{race.name}
						</option>
					))}
				</select>                                    
			</div>

			<div className="CreatorFormIdentity__group align-bottom">
				<h4>Favored Class:</h4>
				<div>
					{race.favoredClass ? jobClasses[race.favoredClass].name : 'Any'}
				</div>
			</div>

			<div className="CreatorFormIdentity__group">
				<h4>Size:</h4>                                 
				<div>{race.size}</div>
			</div> 

			<div className="CreatorFormIdentity__group">
				<h4>Speed:</h4>                                 
				<div>{race.speed}</div>
			</div> 

			<div className="CreatorFormIdentity__group--bottom">
				<h4>Racial Modifiers:</h4>     
				<div className="CreatorForm__input-group">
					{_.filter(race.abilityMods, mod => mod !== 0).length > 0 ?
						_.filter(_.toPairs(race.abilityMods), mod => mod[1] !== 0).map(mod => (
							<div key={mod}>{mod[0].toUpperCase()}: {mod[1]}</div>
						))
						:
						<div>None</div>
					}            
				</div>
			</div>

			<div className="CreatorFormIdentity__group--bottom">
				<h4>Default Language(s):</h4>
				<div>{race.defaultLanguages.join(', ')}</div>
			</div>

			<div className="CreatorFormIdentity__group--top">
				<h4>Bonus Languages:</h4>
				<div>
					{race.bonusLanguages.join(', ')}
				</div>                                    
			</div>

			<div className="CreatorFormIdentity__group--top">
				<h4>Racial Bonuses:</h4>
				{race.racialBonuses ? 
					<ul className="CreatorFormIdentity__list">
						{race.racialBonuses.map((item, i) => (
							<li key={`racialBonus${i}`}>{item}</li>
						))}
					</ul>
					:
					<p>None</p>
				}
			</div>

			<div className="CreatorFormIdentity__group--top">
				<h4>Special Abilities:</h4>
				<div>
					{race.specialAbilities ?
						<SpecialAbilities specialAbilityIds={race.specialAbilities} />
						:
						"None"
					}
				</div>
			</div>        
			
		</div>
	)
}


export default CreatorFormIdentity