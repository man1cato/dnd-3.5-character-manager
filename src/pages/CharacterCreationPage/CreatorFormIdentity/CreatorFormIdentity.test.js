import React from 'react'
import { fireEvent } from '@testing-library/react'

import { apiData, renderWithRedux, FormikWrapper } from '../../../tests/utils'
import CreatorFormIdentity from './CreatorFormIdentity'
import { characterOne } from '../../../tests/seedDatabase'


const setTouched = jest.fn()
const setFieldValue = jest.fn()
let props, state

beforeAll(async () => {
   const api = await apiData()
   props = {
      values: {
         race: api.races[characterOne.race]
      },      
      races: api.races,
      jobClasses: api.jobClasses,
      setTouched,
      setFieldValue
   }
   state = {
      specialAbilities: api.specialAbilities 
   }
})


test('should render CreatorFormIdentity correctly', () => {
   const { container } = renderWithRedux(<CreatorFormIdentity {...props} />, state, FormikWrapper)
   expect(container.firstChild).toMatchSnapshot()
})

