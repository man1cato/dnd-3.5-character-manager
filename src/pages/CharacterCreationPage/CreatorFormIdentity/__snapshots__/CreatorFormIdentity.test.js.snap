// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreatorFormIdentity correctly 1`] = `
<form>
  <div
    class="container--body"
  >
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Name:
      </h4>
      <div>
        <input
          class="text-input"
          name="name"
          value=""
        />
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Gender:
      </h4>
      <select
        class="select"
        name="gender"
      >
        <option
          value="Male"
        >
          Male
        </option>
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Non-Binary"
        >
          Non-Binary
        </option>
      </select>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Age:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="age"
            type="number"
            value=""
          />
          <div>
             yrs
          </div>
        </div>
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Height:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="height.ft"
            type="number"
            value=""
          />
          <div>
            ft
          </div>
          <input
            class="text-input"
            name="height.in"
            type="number"
            value=""
          />
          <div>
            in
          </div>
        </div>
        <div
          class="CreatorForm__input-group"
        />
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Weight:
      </h4>
      <div>
        <div
          class="CreatorForm__input-group"
        >
          <input
            class="text-input"
            name="weight"
            type="number"
            value=""
          />
          <div>
            lbs
          </div>
        </div>
      </div>
    </div>
    <div
      class="divider"
    />
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Race:
      </h4>
      <select
        class="select"
      >
        <option
          selected=""
          value="recOadObhQKOMa7f2"
        >
          Gray Elf
        </option>
        <option
          value="recUDu56yAAHN3Ue3"
        >
          Half-Elf
        </option>
        <option
          value="recWfP0nya01jKP44"
        >
          Half-Orc
        </option>
        <option
          value="recw2GdtKxEQhIpGm"
        >
          High Elf
        </option>
        <option
          value="recv3zpW98noaCctp"
        >
          Hill Dwarf
        </option>
        <option
          value="recxyRHegDBrFYJP5"
        >
          Human
        </option>
        <option
          value="recFEpQ9zuD9Yqu7l"
        >
          Lightfoot Halfling
        </option>
        <option
          value="recrsiTIDAQXD6mCo"
        >
          Mountain Dwarf
        </option>
        <option
          value="recoDD3gAaQXU3A56"
        >
          Orc
        </option>
        <option
          value="recd3zkVkYKZJP92G"
        >
          Rock Gnome
        </option>
        <option
          value="recGQmGZazcjLyEXv"
        >
          Wild Elf
        </option>
        <option
          value="recEPV5rASoteAGoq"
        >
          Wood Elf
        </option>
      </select>
    </div>
    <div
      class="CreatorFormIdentity__group align-bottom"
    >
      <h4>
        Favored Class:
      </h4>
      <div>
        Wizard
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Size:
      </h4>
      <div>
        Medium
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group"
    >
      <h4>
        Speed:
      </h4>
      <div>
        30
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--bottom"
    >
      <h4>
        Racial Modifiers:
      </h4>
      <div
        class="CreatorForm__input-group"
      >
        <div>
          CON
          : 
          -2
        </div>
        <div>
          DEX
          : 
          2
        </div>
        <div>
          INT
          : 
          2
        </div>
        <div>
          STR
          : 
          -2
        </div>
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--bottom"
    >
      <h4>
        Default Language(s):
      </h4>
      <div>
        Common, Elven
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Bonus Languages:
      </h4>
      <div>
        Draconic, Gnoll, Gnome, Goblin, Orc, Sylvan
      </div>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Racial Bonuses:
      </h4>
      <ul
        class="CreatorFormIdentity__list"
      >
        <li>
          Immunity to Sleep effects
        </li>
        <li>
          +2 saving throw bonus against enchantment spells or effects
        </li>
        <li>
          +2 on Listen, Search, and Spot checks. Merely passing within 5 feet of a secret or concealed door entitles a Search check to notice as if actively looking for it.
        </li>
      </ul>
    </div>
    <div
      class="CreatorFormIdentity__group--top"
    >
      <h4>
        Special Abilities:
      </h4>
      <div>
        <div>
          <button
            class="button--link"
            data-testid="recoVjIceMNsVyiqI"
            type="button"
          >
            Low-Light Vision
          </button>
        </div>
      </div>
    </div>
  </div>
</form>
`;
